<a href="https://colab.research.google.com/github/YayraBoven/OAKE-Referral/blob/main/OAKE_Referral_AI.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!git clone https://github.com/YayraBoven/OAKE-Referral.git

Cloning into 'OAKE-Referral'...
remote: Enumerating objects: 16, done.
remote: Counting objects: 100% (16/16), done.
remote: Compressing objects: 100% (15/15), done.
remote: Total 16 (delta 2), reused 0 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (16/16), 34.06 KiB | 1.42 MiB/s, done.
Resolving deltas: 100% (2/2), done.


In [2]:
%cd OAKE-Referral

/content/OAKE-Referral


In [3]:
%%writefile requirements.txt

Writing requirements.txt


In [4]:
!ls

OAKE_Referral_AI.ipynb	README.md  requirements.txt


**Phase O: Data Exploration for Datasets**

In this prep phase, the aim is to understand the columns, data types, length,missing values, and general distribution of the datasets. This is needed to properly carry out our upcoming phase where we encode and clean our three datasets.

In [5]:
import numpy as np
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

sns.set_theme(style="whitegrid")
pd.set_option("display.max_columns", None)
RANDOM_STATE = 42

length_of_stay = pd.read_csv("LengthOfStay.csv")
hosp_inpatient_discharge = pd.read_csv("Hospital_Inpatient_Discharges__SPARCS_De-Identified___2021_20231012.csv")
hospitals = pd.read_csv("hospitals_with_headers.csv")


/tmp/ipykernel_809/2195257636.py:12: DtypeWarning: Columns (29) have mixed types. Specify dtype option on import or set low_memory=False.
  hosp_inpatient_discharge = pd.read_csv("Hospital_Inpatient_Discharges__SPARCS_De-Identified___2021_20231012.csv")


In [6]:
#Inspecting length_of_stay
# Displays the names of all the columns in the dataset.
print("COLUMN NAMES")
print(length_of_stay.columns)

# Returns the number of rows and columns in the dataset.
print("DATASET SHAPES")
print(length_of_stay.shape)


# Displays the first five rows to give an overview of the dataset.
print("FIRST FIVE ROWS")

print(length_of_stay.head())

# Shows information about the dataset, including data types, non-null values, and memory usage.
print("DATASET INFORMATION")
length_of_stay.info()

# Returns summary statistics for all numerical columns in the dataset.
print("SUMMARY STATISTICS")
print(length_of_stay.describe())

# Counts the number of missing values in each column.
print("MISSING VALUES")
print(length_of_stay.isna().sum())

# Counts the total number of duplicate rows in the dataset.
print("DUPLICATE ROWS")
print(length_of_stay.duplicated().sum())

# Displays the data type of each column.
print("DATA TYPES")
print(length_of_stay.dtypes)



COLUMN NAMES
Index(['eid', 'vdate', 'rcount', 'gender', 'dialysisrenalendstage', 'asthma',
       'irondef', 'pneum', 'substancedependence', 'psychologicaldisordermajor',
       'depress', 'psychother', 'fibrosisandother', 'malnutrition', 'hemo',
       'hematocrit', 'neutrophils', 'sodium', 'glucose', 'bloodureanitro',
       'creatinine', 'bmi', 'pulse', 'respiration',
       'secondarydiagnosisnonicd9', 'discharged', 'facid', 'lengthofstay'],
      dtype='object')
DATASET SHAPES
(100000, 28)
FIRST FIVE ROWS
   eid       vdate rcount gender  dialysisrenalendstage  asthma  irondef  \
0    1   8/29/2012      0      F                      0       0        0   
1    2   5/26/2012     5+      F                      0       0        0   
2    3   9/22/2012      1      F                      0       0        0   
3    4    8/9/2012      0      F                      0       0        0   
4    5  12/20/2012      0      F                      0       0        0   

   pneum  substancedependen

In [7]:
#Inspecting hosp_inpatient_discharge
# Displays the names of all the columns in the dataset.
print("COLUMN NAMES")
print(hosp_inpatient_discharge.columns)

# Returns the number of rows and columns in the dataset.
print("DATASET SHAPES")
print(hosp_inpatient_discharge.shape)


# Displays the first five rows to give an overview of the dataset.
print("FIRST FIVE ROWS")

print(hosp_inpatient_discharge.head())

# Shows information about the dataset, including data types, non-null values, and memory usage.
print("DATASET INFORMATION")
hosp_inpatient_discharge.info()

# Returns summary statistics for all numerical columns in the dataset.
print("SUMMARY STATISTICS")
print(hosp_inpatient_discharge.describe())

# Counts the number of missing values in each column.
print("MISSING VALUES")
print(hosp_inpatient_discharge.isna().sum())

# Counts the total number of duplicate rows in the dataset.
print("DUPLICATE ROWS")
print(hosp_inpatient_discharge.duplicated().sum())

# Displays the data type of each column.
print("DATA TYPES")
print(hosp_inpatient_discharge.dtypes)

COLUMN NAMES
Index(['Hospital Service Area', 'Hospital County',
       'Operating Certificate Number', 'Permanent Facility Id',
       'Facility Name', 'Age Group', 'Zip Code - 3 digits', 'Gender', 'Race',
       'Ethnicity', 'Length of Stay', 'Type of Admission',
       'Patient Disposition', 'Discharge Year', 'CCSR Diagnosis Code',
       'CCSR Diagnosis Description', 'CCSR Procedure Code',
       'CCSR Procedure Description', 'APR DRG Code', 'APR DRG Description',
       'APR MDC Code', 'APR MDC Description', 'APR Severity of Illness Code',
       'APR Severity of Illness Description', 'APR Risk of Mortality',
       'APR Medical Surgical Description', 'Payment Typology 1',
       'Payment Typology 2', 'Payment Typology 3', 'Birth Weight',
       'Emergency Department Indicator', 'Total Charges', 'Total Costs'],
      dtype='object')
DATASET SHAPES
(402756, 33)
FIRST FIVE ROWS
  Hospital Service Area Hospital County  Operating Certificate Number  \
0         New York City           

In [8]:
print(hosp_inpatient_discharge['Patient Disposition'].value_counts())

print(hosp_inpatient_discharge['APR Severity of Illness Description'].value_counts())

print(hosp_inpatient_discharge['APR Risk of Mortality'].value_counts())

print(hosp_inpatient_discharge['Type of Admission'].value_counts())

print(hosp_inpatient_discharge['APR Medical Surgical Description'].value_counts())

print(hosp_inpatient_discharge['Length of Stay'].value_counts().head(30))

Patient Disposition
Home or Self Care                        255108
Home w/ Home Health Services              62890
Skilled Nursing Home                      32863
Expired                                   15187
Left Against Medical Advice               11094
Short-term Hospital                        6752
Inpatient Rehabilitation Facility          6626
Hospice - Home                             2611
Another Type Not Listed                    2112
Psychiatric Hospital or Unit of Hosp       2032
Hospice - Medical Facility                 1919
Facility w/ Custodial/Supportive Care       930
Court/Law Enforcement                       851
Hosp Basd Medicare Approved Swing Bed       631
Medicare Cert Long Term Care Hospital       544
Cancer Center or Children's Hospital        276
Medicaid Cert Nursing Facility              191
Federal Health Care Facility                102
Critical Access Hospital                     37
Name: count, dtype: int64
APR Severity of Illness Description
Modera

In [9]:
#Inspecting hospitals
# Displays the names of all the columns in the dataset.
print("COLUMN NAMES")
print(hospitals.columns)

# Returns the number of rows and columns in the dataset.
print("DATASET SHAPES")
print(hospitals.shape)


# Displays the first five rows to give an overview of the dataset.
print("FIRST FIVE ROWS")

print(hospitals.head())

# Shows information about the dataset, including data types, non-null values, and memory usage.
print("DATASET INFORMATION")
hospitals.info()

# Returns summary statistics for all numerical columns in the dataset.
print("SUMMARY STATISTICS")
print(hospitals.describe())

# Counts the number of missing values in each column.
print("MISSING VALUES")
print(hospitals.isna().sum())

# Counts the total number of duplicate rows in the dataset.
print("DUPLICATE ROWS")
print(hospitals.duplicated().sum())

# Displays the data type of each column.
print("DATA TYPES")
print(hospitals.dtypes)

COLUMN NAMES
Index(['hospital_id', 'hospital_name', 'region', 'location', 'hospital_level',
       'specialities', 'number_of_beds', 'beds_occupied', 'beds_available'],
      dtype='object')
DATASET SHAPES
(24, 9)
FIRST FIVE ROWS
   hospital_id                                     hospital_name  \
0            1                        Korle Bu Teaching Hospital   
1            2                University of Ghana Medical Centre   
2            3                              37 Military Hospital   
3            4                             Tema General Hospital   
4            5  Ridge Hospital (Greater Accra Regional Hospital)   

          region                                           location  \
0  Greater Accra  Guggisberg Avenue, Ablekuma South District, Accra   
1  Greater Accra           University of Ghana Campus, Legon, Accra   
2  Greater Accra                             Liberation Road, Accra   
3  Greater Accra                   Hospital Road, Community 9, Tema   
4  Gre

In [10]:
#Here, we remove Unnecessary Columns
length_of_stay = length_of_stay.drop(columns=['eid', 'vdate', 'discharged'])

#Updated Columns
print("Updated Columns")
print(length_of_stay.columns)

Updated Columns
Index(['rcount', 'gender', 'dialysisrenalendstage', 'asthma', 'irondef',
       'pneum', 'substancedependence', 'psychologicaldisordermajor', 'depress',
       'psychother', 'fibrosisandother', 'malnutrition', 'hemo', 'hematocrit',
       'neutrophils', 'sodium', 'glucose', 'bloodureanitro', 'creatinine',
       'bmi', 'pulse', 'respiration', 'secondarydiagnosisnonicd9', 'facid',
       'lengthofstay'],
      dtype='object')


In [11]:
print(length_of_stay.select_dtypes(include=["object"]).columns)

Index(['rcount', 'gender', 'facid'], dtype='object')


In [12]:
print(length_of_stay["rcount"].head(10))

0     0
1    5+
2     1
3     0
4     0
5     3
6     4
7     0
8     0
9     0
Name: rcount, dtype: object


**Phase 1 : Encoding & Final Cleaning**

After assessing the information
about our three datasets, we start cleaning and encoding them.

In this phase, we do not alter the APR Severity of Illness Description, APR Risk of Mortality and Patient disposition columns of the Hospital Inpatient Discharge Dataset.
This is because the original values are needed in our Phase 2 to create our suitable_for_downward_referral target.



In [13]:
#Encoding rcount, gender and facid because they are categorical and the model needs to work numerically.
from sklearn.preprocessing import LabelEncoder

encoder = LabelEncoder()

categorical_columns = ["gender", "facid"]

for col in categorical_columns:
    length_of_stay[col] = encoder.fit_transform(length_of_stay[col])


# We replace "5+" (patients with 5 or more previous counts) with 5
length_of_stay["rcount"] = length_of_stay["rcount"].replace("5+", 5)

# We also convert rcount from object values to integers
length_of_stay["rcount"] = length_of_stay["rcount"].astype(int)

In [14]:
#We create a copy, to prevent altering the original dataset.
hosp_discharge_df = hosp_inpatient_discharge.copy()

# Here, duplicate records are removed to produce optimal performance in the model.
hosp_discharge_df = hosp_discharge_df.drop_duplicates()

# We also drop columns that are not useful for the referral model such as administrative columns.
#Currently, our model focuses on informayion relating to the patient's clinical information, treatment etc.
#Unrelated columns could add noise and cause the model to learn hospital-specific patterns.
columns_to_drop = [
    'Hospital Service Area',
    'Hospital County',
    'Operating Certificate Number',
    'Permanent Facility Id',
    'Facility Name',
    'Zip Code - 3 digits',
    'Discharge Year',
    'Payment Typology 1',
    'Payment Typology 2',
    'Payment Typology 3',
    'Birth Weight',
    'Total Charges',
    'Total Costs',
    'Race',
    'Ethnicity'
]

hosp_discharge_df = hosp_discharge_df.drop(columns=columns_to_drop)

In [15]:
import os
#The length of stay values are converted from objects to numbers and invalid values are changed to NaN  using errors = ;coerce'
hosp_discharge_df['Length of Stay'] = pd.to_numeric(
    hosp_discharge_df['Length of Stay'],
    errors='coerce'
)
# We store the APR code columns as nullable integers, which allow missing values.
hosp_discharge_df['APR DRG Code'] = hosp_discharge_df['APR DRG Code'].astype('Int64')
hosp_discharge_df['APR MDC Code'] = hosp_discharge_df['APR MDC Code'].astype('Int64')
hosp_discharge_df['APR Severity of Illness Code'] = hosp_discharge_df['APR Severity of Illness Code'].astype('Int64')

#Here we identify the categorical columns and make sure that spaces don't prevent the identification of the same value
categorical_columns = hosp_discharge_df.select_dtypes(
    include='object'
).columns

for col in categorical_columns:
    hosp_discharge_df[col] = hosp_discharge_df[col].str.strip()


#This part removes records with no diagnosis
hosp_discharge_df = hosp_discharge_df.dropna(subset=['CCSR Diagnosis Code', 'CCSR Diagnosis Description'])
#This part removes records with no lenth of stay value
hosp_discharge_df = hosp_discharge_df.dropna(subset=['Length of Stay'])

#Here, we treat missing procedures as no recorded procedure, using "None"
hosp_discharge_df['CCSR Procedure Code'] = (hosp_discharge_df['CCSR Procedure Code'].fillna('None'))
hosp_discharge_df['CCSR Procedure Description'] = (hosp_discharge_df['CCSR Procedure Description'].fillna('None'))


#We remove missing severity and mortality risk values due to their importance
hosp_discharge_df = hosp_discharge_df.dropna(subset=['APR Severity of Illness Description', 'APR Risk of Mortality'])


os.makedirs('data/processed', exist_ok=True)

#We save the cleaned dataset to prevent extra pre-processing
hosp_discharge_df.to_csv('data/processed/hospital_discharge_cleaned.csv',index=False)

#check the result
print(hosp_discharge_df.shape)

print(hosp_discharge_df.isnull().sum())

print(hosp_discharge_df.dtypes)


(402013, 18)
Age Group                              0
Gender                                 0
Length of Stay                         0
Type of Admission                      0
Patient Disposition                    0
CCSR Diagnosis Code                    0
CCSR Diagnosis Description             0
CCSR Procedure Code                    0
CCSR Procedure Description             0
APR DRG Code                           0
APR DRG Description                    0
APR MDC Code                           0
APR MDC Description                    0
APR Severity of Illness Code           0
APR Severity of Illness Description    0
APR Risk of Mortality                  0
APR Medical Surgical Description       0
Emergency Department Indicator         0
dtype: int64
Age Group                               object
Gender                                  object
Length of Stay                         float64
Type of Admission                       object
Patient Disposition                     object
C

In [16]:
# In this cell, we mainly drop redundant or description columns.
hosp_discharge_df = hosp_discharge_df.drop(columns=[
    'CCSR Diagnosis Description',
    'CCSR Procedure Description',
    'APR DRG Description',
    'APR MDC Description'
])

print(hosp_discharge_df.shape)


(402013, 14)


In [17]:
# We use LabelEncoder to encode remaining categorical columns so they are useful to our model
from sklearn.preprocessing import LabelEncoder

# For Age Groups , we assign values that follow the natiral progression from young to old
age_order = {'0 to 17': 0, '18 to 29': 1, '30 to 49': 2, '50 to 69': 3, '70 or Older': 4}
hosp_discharge_df['Age Group'] = hosp_discharge_df['Age Group'].map(age_order)

# For Gender, we use ne-hot (M/F/U, only 3 categories)
hosp_discharge_df = pd.get_dummies(hosp_discharge_df, columns=['Gender'], prefix='Gender')

# Type of Admission -> one-hot (6 categories, no natural order)
hosp_discharge_df = pd.get_dummies(hosp_discharge_df, columns=['Type of Admission'], prefix='Admission')

# CCSR Diagnosis / Procedure Code -> label encode (too high-cardinality for one-hot: 477 / 321 categories)
le_diag = LabelEncoder()
hosp_discharge_df['CCSR Diagnosis Code'] = le_diag.fit_transform(hosp_discharge_df['CCSR Diagnosis Code'])

le_proc = LabelEncoder()
hosp_discharge_df['CCSR Procedure Code'] = le_proc.fit_transform(hosp_discharge_df['CCSR Procedure Code'])

# APR Medical Surgical Description -> binary
hosp_discharge_df['APR Medical Surgical Description'] = hosp_discharge_df['APR Medical Surgical Description'].map(
    {'Medical': 0, 'Surgical': 1}
)

# Emergency Department Indicator -> binary
hosp_discharge_df['Emergency Department Indicator'] = hosp_discharge_df['Emergency Department Indicator'].map(
    {'N': 0, 'Y': 1}
)

print(hosp_discharge_df.shape)
print(hosp_discharge_df.dtypes)


(402013, 21)
Age Group                                int64
Length of Stay                         float64
Patient Disposition                     object
CCSR Diagnosis Code                      int64
CCSR Procedure Code                      int64
APR DRG Code                             Int64
APR MDC Code                             Int64
APR Severity of Illness Code             Int64
APR Severity of Illness Description     object
APR Risk of Mortality                   object
APR Medical Surgical Description         int64
Emergency Department Indicator           int64
Gender_F                                  bool
Gender_M                                  bool
Gender_U                                  bool
Admission_Elective                        bool
Admission_Emergency                       bool
Admission_Newborn                         bool
Admission_Not Available                   bool
Admission_Trauma                          bool
Admission_Urgent                          bool


In [18]:
#We save the cleaned and encoded hospital discharge dataset to be reused
import os
os.makedirs('data/processed', exist_ok=True)
hosp_discharge_df.to_csv('data/processed/hosp_discharge_phase1.csv', index=False)
print("Saved data/processed/hosp_discharge_phase1.csv")


Saved data/processed/hosp_discharge_phase1.csv


In [19]:
# We decided to drop the facid column, because it is just used to identify a facility and could lead to our model learning facility-biased patterns
length_of_stay = length_of_stay.drop(columns=['facid'])
print(length_of_stay.shape)
print(length_of_stay.columns.tolist())


(100000, 24)
['rcount', 'gender', 'dialysisrenalendstage', 'asthma', 'irondef', 'pneum', 'substancedependence', 'psychologicaldisordermajor', 'depress', 'psychother', 'fibrosisandother', 'malnutrition', 'hemo', 'hematocrit', 'neutrophils', 'sodium', 'glucose', 'bloodureanitro', 'creatinine', 'bmi', 'pulse', 'respiration', 'secondarydiagnosisnonicd9', 'lengthofstay']


In [20]:
# A Data quality check is performed.
#We confirm that the following conditions are represented as binary features before we use the datasets.
binary_flags = ['dialysisrenalendstage','asthma','irondef','pneum','substancedependence',
                 'psychologicaldisordermajor','depress','psychother','fibrosisandother',
                 'malnutrition','hemo']
for c in binary_flags:
    print(c, length_of_stay[c].unique())


dialysisrenalendstage [0 1]
asthma [0 1]
irondef [0 1]
pneum [0 1]
substancedependence [0 1]
psychologicaldisordermajor [0 1]
depress [0 1]
psychother [0 1]
fibrosisandother [0 1]
malnutrition [0 1]
hemo [0 1]


In [ ]:
# Another data quality check happens here,
#We verify labs/vitals for any impossible values
labs = ['hematocrit','neutrophils','sodium','glucose','bloodureanitro','creatinine','bmi','pulse','respiration']
print(length_of_stay[labs].describe())

# 1 row with negative glucose is found and so it is dropped
print("\nNegative glucose rows:", (length_of_stay['glucose'] < 0).sum())
length_of_stay = length_of_stay[length_of_stay['glucose'] >= 0]
print("Shape after dropping:", length_of_stay.shape)

# We keep rare extreme values in neutrophils/bloodureanitro (~0.5% of rows).
# They serve as a clinically meaningful severity signal for a referral model, not noise to be removed.


          hematocrit    neutrophils         sodium        glucose  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean       11.975939      10.177455     137.891397     141.963384   
std         2.031792       5.353131       2.999669      29.992996   
min         4.400000       0.100000     124.912632      -1.005927   
25%        10.900000       7.700000     135.871062     121.682383   
50%        11.900000       9.400000     137.887151     142.088545   
75%        12.900000      11.500000     139.912885     162.180996   
max        24.100000     245.900000     151.387283     271.444277   

       bloodureanitro     creatinine            bmi          pulse  \
count   100000.000000  100000.000000  100000.000000  100000.000000   
mean        14.097185       1.099350      29.805759      73.444720   
std         12.952454       0.200262       2.003769      11.644555   
min          1.000000       0.219770      21.992683      21.000000   
25%         11.000000       

In [ ]:
length_of_stay.to_csv('data/processed/length_of_stay_phase1.csv', index=False)
print("Saved data/processed/length_of_stay_phase1.csv")


Saved data/processed/length_of_stay_phase1.csv


In [21]:
# For the hospitals_with_headers dataset, we do not need to clean it. This is because it has no nulls
#However, we would add a hospital_tier column later on to simplify the 14 text variants of hospital_level.
print(hospitals.isnull().sum())
print(hospitals['hospital_level'].unique())


hospital_id       0
hospital_name     0
region            0
location          0
hospital_level    0
specialities      0
number_of_beds    0
beds_occupied     0
beds_available    0
dtype: int64
['Teaching Hospital' 'Academic Medical Centre (Quaternary)'
 'District Hospital' 'Regional Hospital' 'Primary/District Hospital'
 'Primary Hospital' 'Primary Hospital (Private)'
 'Primary Hospital (Mission)' 'Secondary Hospital'
 'District/Municipal Hospital' 'Municipal Hospital'
 'Municipal/General Hospital' 'Community Hospital'
 'Primary/Public Hospital']


In [22]:
# This function simplifies the hospital-level descriptions into 3 tiers
#For easier recommendations and identification
def map_tier(level):
    level = level.lower()
    if 'teaching' in level or 'quaternary' in level or 'regional' in level:
        return 1  # Major referral-level
    elif 'district' in level or 'municipal' in level or 'secondary' in level:
        return 2  # Mid-level
    else:
        return 3  # Primary/Community

hospitals['hospital_tier'] = hospitals['hospital_level'].apply(map_tier)
print(hospitals[['hospital_name', 'hospital_level', 'hospital_tier']])

hospitals.to_csv('data/processed/hospitals_phase1.csv', index=False)
print("Saved data/processed/hospitals_phase1.csv")


                                        hospital_name  \
0                          Korle Bu Teaching Hospital   
1                  University of Ghana Medical Centre   
2                                37 Military Hospital   
3                               Tema General Hospital   
4    Ridge Hospital (Greater Accra Regional Hospital)   
5                          Ga East Municipal Hospital   
6                                   Achimota Hospital   
7                         Ga North Municipal Hospital   
8                                North Legon Hospital   
9                          Pentecost Hospital, Madina   
10                                    Police Hospital   
11     Ga West Municipal Hospital (Amasaman Hospital)   
12                   Lapaz Community Hospital Annex A   
13     Ghana Atomic Energy Commission (GAEC) Hospital   
14                                     LEKMA Hospital   
15                     Weija-Gbawe Municipal Hospital   
16                           Ma

**Phase 2: Target Variable Engineering**

Patient Disposition is used to create our target because it is the closest to an outcome.
APR Severity of Illness and APR Risk of Mortality are clinical assessments (available at time of care), so they become input features here, not part of the label. This avoids the model leaking its own answer.

**suitable_for_downward_referral = 1**
Stepped down to a lower-level hospital.

**suitable_for_downward_referral = 0**
Still in need of  continued high-level/specialized care

**Excluded from labeled data** Administrative/legal, not a clinical stability judgment

In [23]:
#These are outcomes that represent a downward referral for less-severe care
positive = ['Home or Self Care', 'Home w/ Home Health Services', 'Skilled Nursing Home',
            'Inpatient Rehabilitation Facility', 'Facility w/ Custodial/Supportive Care',
            'Medicare Cert Long Term Care Hospital', 'Hosp Basd Medicare Approved Swing Bed',
            'Medicaid Cert Nursing Facility', 'Critical Access Hospital']
#These are the outcomes that suggest a patient cannot be downward-referred and still needs specialised care.
negative = ['Expired', 'Short-term Hospital', 'Psychiatric Hospital or Unit of Hosp',
            'Hospice - Home', 'Hospice - Medical Facility', "Cancer Center or Children's Hospital"]

#These do not really show any information concerning patient stability
excluded = ['Left Against Medical Advice', 'Court/Law Enforcement',
            'Another Type Not Listed', 'Federal Health Care Facility']
#We drop these columns here
hosp_discharge_df = hosp_discharge_df[~hosp_discharge_df['Patient Disposition'].isin(excluded)]

#Here, a binary target is created where 1 means suitable for downward referral and 0 means not suitable
hosp_discharge_df['suitable_for_downward_referral'] = hosp_discharge_df['Patient Disposition'].apply(
    lambda x: 1 if x in positive else 0
)
#These 2 lines check the number and percentage in the target classes.
print(hosp_discharge_df['suitable_for_downward_referral'].value_counts())
print(hosp_discharge_df['suitable_for_downward_referral'].value_counts(normalize=True).round(4) * 100)


suitable_for_downward_referral
1    359170
0     28709
Name: count, dtype: int64
suitable_for_downward_referral
1    92.6
0     7.4
Name: proportion, dtype: float64


The target is imbalanced with about 92.6% referrable and 7.4% not referrable. This reflects real clinical situations where most admissions end in routine discharge.

Due to this we may need class weighting and evaluate with precision/recall/F1.

In [24]:
# We encode Severity of Illness and Risk of Mortality ordinally, due to the higher progression in them.
severity_order = {'Minor': 0, 'Moderate': 1, 'Major': 2, 'Extreme': 3}
hosp_discharge_df['APR Severity of Illness Description'] = hosp_discharge_df['APR Severity of Illness Description'].map(severity_order)

risk_order = {'Minor': 0, 'Moderate': 1, 'Major': 2, 'Extreme': 3}
hosp_discharge_df['APR Risk of Mortality'] = hosp_discharge_df['APR Risk of Mortality'].map(risk_order)
#We rename the columns to properly show the encoded values
hosp_discharge_df = hosp_discharge_df.rename(columns={'APR Risk of Mortality': 'APR Risk of Mortality Code'})

#The APR Severity of Illness Code is now redundant with the Description we just encoded, so we drop duplicates
hosp_discharge_df = hosp_discharge_df.drop(columns=['APR Severity of Illness Description'])

# Patient Disposition was used to creaet the target, so keeping it as a feature would be a leakage, so it is dropped.
hosp_discharge_df = hosp_discharge_df.drop(columns=['Patient Disposition'])

print(hosp_discharge_df.shape)
print(hosp_discharge_df.columns.tolist())
print("Nulls:", hosp_discharge_df.isnull().sum().sum())


(387879, 20)
['Age Group', 'Length of Stay', 'CCSR Diagnosis Code', 'CCSR Procedure Code', 'APR DRG Code', 'APR MDC Code', 'APR Severity of Illness Code', 'APR Risk of Mortality Code', 'APR Medical Surgical Description', 'Emergency Department Indicator', 'Gender_F', 'Gender_M', 'Gender_U', 'Admission_Elective', 'Admission_Emergency', 'Admission_Newborn', 'Admission_Not Available', 'Admission_Trauma', 'Admission_Urgent', 'suitable_for_downward_referral']
Nulls: 0


In [26]:
hosp_discharge_df.to_csv('data/processed/hosp_discharge_phase2_final.csv', index=False)
print("Saved data/processed/hosp_discharge_phase2_final.csv - ready for Phase 3 modeling")


Saved data/processed/hosp_discharge_phase2_final.csv - ready for Phase 3 modeling


length_of_stay role for Phase 3

This dataset does not have a column that works like a disposition outcome. There is nothing similar to Patient Disposition that could be used to create a comparable label. If we forced an artificial proxy, like setting a cutoff point on lengthofstay, it would add weak and unreliable labels into the training process.
Decision: hosp_discharge_df, which comes from SPARCS, will be used as the main training set for the classifier, since it contains real outcome data that the model can learn from.length_of_stay will be treated as a secondary and exploratory resource. It is useful for checking how comorbidities and lab values relate to length of stay, but it will not be used to train the target variable for the main classifier.


length_of_stay role for Phase 3

This dataset does not have a column that works like a disposition outcome. There is nothing similar to Patient Disposition that could be used to create a comparable label. If we forced an artificial proxy, like setting a cutoff point on lengthofstay, it would add weak and unreliable labels into the training process.
Decision: hosp_discharge_df, which comes from SPARCS, will be used as the main training set for the classifier, since it contains real outcome data that the model can learn from.length_of_stay will be treated as a secondary and exploratory resource. It is useful for checking how comorbidities and lab values relate to length of stay, but it will not be used to train the target variable for the main classifier.


**Phase 3: Training & Evaluation of Referral Classifier**

For training, We did a stratified 80/20(80 for train and 20 for split) split because we had a 93.5/6.5 class imbalance from phase 2.We use 'class_weight='balanced'' on both models so the rare 'not suitable' class isn't ignored.Without it , the model could just predict 'referrable' every time and score ~93.5% accuracy while being clinically useless.

Our evaluation focuses on precision/recall/F1.
Class 0 ('Not suitable') is a critical error to keenly watch because it measures how often the model correctly catches patients who should NOT be referred down.When this is missed , it creates a false negative on class 0 which is a dangerous failure mode because it would lead to clearing someone who actually needs continued major-hospital care.




In [27]:
#Import train_test_split for splitting data into train and test sets
from sklearn.model_selection import train_test_split

#Separate features (X) and target (y = suitable_for_downward_referral)
X = hosp_discharge_df.drop(columns=['suitable_for_downward_referral'])
y = hosp_discharge_df['suitable_for_downward_referral']

#Split into train (80%) and test (20%) sets
#stratify=y ensures both sets have the same proportion of each class as the original data
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

#Check the shapes to confirm the split worked correctly
print(f"Train: {X_train.shape}, Test: {X_test.shape}")

#Check the class balance in the training set to confirm stratification worked
print("Train class balance:\n", y_train.value_counts(normalize=True).round(4))


Train: (310303, 19), Test: (77576, 19)
Train class balance:
 suitable_for_downward_referral
1    0.926
0    0.074
Name: proportion, dtype: float64


In [28]:
#Import Decision Tree classifier and evaluation metrics
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report, confusion_matrix

#Initialise a Decision Tree as a quick, interpretable baseline model
#class_weight='balanced' adjusts for any class imbalance in the target
#max_depth=10 limits the tree depth to help prevent overfitting
dt = DecisionTreeClassifier(random_state=42, class_weight='balanced', max_depth=10)
dt.fit(X_train, y_train)

#Predict on the test set
y_pred_dt = dt.predict(X_test)

#Print precision, recall, f1-score, and support for both classes
print("Decision Tree (baseline)")
print(classification_report(y_test, y_pred_dt, target_names=['Not suitable (0)', 'Suitable (1)']))

#Print the confusion matrix to see the breakdown of correct vs incorrect predictions per class
print("Confusion matrix:")
print(confusion_matrix(y_test, y_pred_dt))


Decision Tree (baseline)
                  precision    recall  f1-score   support

Not suitable (0)       0.27      0.77      0.40      5742
    Suitable (1)       0.98      0.84      0.90     71834

        accuracy                           0.83     77576
       macro avg       0.62      0.80      0.65     77576
    weighted avg       0.93      0.83      0.86     77576

Confusion matrix:
[[ 4401  1341]
 [11801 60033]]


In [29]:
#Import Random Forest classifier which is  an ensemble of decision trees, usually stronger than a single tree
from sklearn.ensemble import RandomForestClassifier

#Initialise the Random Forest
#n_estimators=50 builds 50 trees; max_depth=12 limits each tree's depth to reduce overfitting
#class_weight='balanced' adjusts for class imbalance; n_jobs=-1 uses all CPU cores to train faster
rf = RandomForestClassifier(n_estimators=50, max_depth=12, random_state=42,
                             class_weight='balanced', n_jobs=-1)
rf.fit(X_train, y_train)

#Predict on the test set
y_pred_rf = rf.predict(X_test)

#Print precision, recall, f1-score, and support for both classes
print("Random Forest")
print(classification_report(y_test, y_pred_rf, target_names=['Not suitable (0)', 'Suitable (1)']))

#Print the confusion matrix to compare against the Decision Tree baseline
print("Confusion matrix:")
print(confusion_matrix(y_test, y_pred_rf))

Random Forest
                  precision    recall  f1-score   support

Not suitable (0)       0.29      0.76      0.42      5742
    Suitable (1)       0.98      0.85      0.91     71834

        accuracy                           0.85     77576
       macro avg       0.64      0.81      0.67     77576
    weighted avg       0.93      0.85      0.88     77576

Confusion matrix:
[[ 4379  1363]
 [10545 61289]]


In [30]:
# Feature importance checks that the model is using clinically sensible signals
importances = pd.Series(rf.feature_importances_, index=X_train.columns).sort_values(ascending=False)
print(importances.round(4))


APR Risk of Mortality Code          0.2982
APR Severity of Illness Code        0.2348
APR DRG Code                        0.0787
CCSR Diagnosis Code                 0.0752
Length of Stay                      0.0704
Age Group                           0.0648
APR MDC Code                        0.0501
CCSR Procedure Code                 0.0415
APR Medical Surgical Description    0.0345
Admission_Emergency                 0.0163
Emergency Department Indicator      0.0136
Admission_Elective                  0.0090
Admission_Newborn                   0.0038
Gender_M                            0.0031
Gender_F                            0.0031
Admission_Urgent                    0.0024
Admission_Trauma                    0.0004
Admission_Not Available             0.0001
Gender_U                            0.0000
dtype: float64


Result: APR Risk of Mortality (30 percent) and APR Severity of Illness (22 percent) are the top two factors driving the model's decisions. This matches what we would expect from a clinical point of view. Diagnosis and DRG codes, length of stay, and age come after these two. Gender and admission type make up less than 1 percent combined, which is reassuring for the fairness discussion in Phase 6.

In [31]:
#save the trained random forest model to disk so it can be reused later without retraining
import joblib
import os

#create a models folder if it does not already exist
os.makedirs('models', exist_ok=True)

#save the model as a joblib file inside the models folder
joblib.dump(rf, 'models/referral_classifier_rf.joblib')
print("Saved models/referral_classifier_rf.joblib")

Saved models/referral_classifier_rf.joblib


Next steps

-Tune the classification threshold to be more conservative, to reduce the 19 to 20 percent miss rate on class 0, which is the safety critical error. Being overly cautious is a much safer outcome than being under cautious in this case.

-Try more hyperparameter tuning, such as using more trees or changing max_depth and min_samples_leaf.

-Use formal cross validation instead of a single train and test split.

-Start building the rule based hospital recommendation system, which is Phase 4.

Demo: Using the saved classifier on new patients

The model has been trained and saved. This section shows what it looks like to actually use it. We create two hypothetical patients that contrast with each other, and load the model fresh from disk, the same way a  the recommendation system would, to confirm it behaves sensibly rather than just relying on objects still sitting in memory.

In [32]:
import joblib

# Load the saved model fresh from disk
model = joblib.load('models/referral_classifier_rf.joblib')

feature_cols = X_train.columns.tolist()

#Patient A: young adult, minor severity, low risk, elective admission, short stay
patient_a = pd.DataFrame([{
    'Age Group': 1, 'Length of Stay': 2.0, 'CCSR Diagnosis Code': 50, 'CCSR Procedure Code': 100,
    'APR DRG Code': 300, 'APR MDC Code': 8, 'APR Severity of Illness Code': 1, 'APR Risk of Mortality Code': 0,
    'APR Medical Surgical Description': 1, 'Emergency Department Indicator': 0,
    'Gender_F': True, 'Gender_M': False, 'Gender_U': False,
    'Admission_Elective': True, 'Admission_Emergency': False, 'Admission_Newborn': False,
    'Admission_Not Available': False, 'Admission_Trauma': False, 'Admission_Urgent': False
}])[feature_cols]

#Patient B: elderly, extreme severity, extreme mortality risk, emergency admission, long stay
patient_b = pd.DataFrame([{
    'Age Group': 4, 'Length of Stay': 21.0, 'CCSR Diagnosis Code': 12, 'CCSR Procedure Code': 5,
    'APR DRG Code': 137, 'APR MDC Code': 4, 'APR Severity of Illness Code': 4, 'APR Risk of Mortality Code': 3,
    'APR Medical Surgical Description': 0, 'Emergency Department Indicator': 1,
    'Gender_F': False, 'Gender_M': True, 'Gender_U': False,
    'Admission_Elective': False, 'Admission_Emergency': True, 'Admission_Newborn': False,
    'Admission_Not Available': False, 'Admission_Trauma': False, 'Admission_Urgent': False
}])[feature_cols]

# run both patients through the model and print the prediction with confidence level
for name, p in [("Patient A (young, minor severity, elective)", patient_a),
                ("Patient B (elderly, extreme severity, emergency)", patient_b)]:
    pred = model.predict(p)[0]
    proba = model.predict_proba(p)[0]
    label = "SUITABLE for downward referral" if pred == 1 else "NOT suitable -- needs continued major-hospital care"
    print(f"{name}")
    print(f"  Prediction: {label}")
    print(f"  Confidence: {proba[pred]:.1%}  (P(not suitable)={proba[0]:.3f}, P(suitable)={proba[1]:.3f})")
    print()


Patient A (young, minor severity, elective)
  Prediction: SUITABLE for downward referral
  Confidence: 96.0%  (P(not suitable)=0.040, P(suitable)=0.960)

Patient B (elderly, extreme severity, emergency)
  Prediction: NOT suitable -- needs continued major-hospital care
  Confidence: 62.9%  (P(not suitable)=0.629, P(suitable)=0.371)



Phase 3 Result: Patient A, who is young with minor severity and an elective admission, is classified as suitable for downward referral with 98.2 percent confidence. Patient B, who is elderly with extreme severity, extreme mortality risk, and an emergency admission, is classified as not suitable with 82.6 percent confidence. This is the output the classifier gives, a decision along with a confidence score, and it is what Phase 4's recommendation system will use for patients who come back with a result of 1, meaning suitable.